### PR TITLE
Improve and standardize input methods

### PR DESCRIPTION
**Pull request by ericsagnes**
Work in progress, please do not merge yet.
implementation of #11138 
### fcitx:
- [x] add nix module
- [x] fcitx-mozc: init at 2.17.2105.102
- [x] fcitx-hangul: init at 0.3.0
- [x] fcitx-chewing: init at 0.2.2
- [x] fcitx-table-other: init at 0.2.3
- [x] fcitx-m17n: init at 0.2.3
- [x] fcitx-qt5: init at 1.0.4
- [x] fcitx: 4.2.8.5 -> 4.2.9
- [x] group engines
### ibus:
- [x] group engines
- [x] refactor nix module
- [x] add ibus-hangul
- [x] add ibus-m17n
- [x] fix ibus-mozc licence
### Librairies:
- [x] ~~libchewing: init at 0.4.0~~ Cherry picked in master
### nabi:
- [x] add nix module
### uim:
- [x] add nix module
### Documentation:
- [x] add documentation for input methods
## Current issues
- ~~fcitx-anthy: any touch beside letters and numbers is inactive.~~ -> Fixed by upgrading `ibus-anthy` to 0.2.2.
- fcitx-configtool: fails to start fcitx-config-gtk3
- ~~fcitx-config-gtk3: "settings" button has no effect~~ -> Fixed by using `ibus-wrapper`
## Current packaging issues
- ibus-hangul:
  - ~~`ibus-hangul-setup tool` is giving runtime errors~~ Fixed by the use of `ibus-wrapper`
  - ~~toolbar icon missing, seems common to all ibus engines. Need more investigation.~~ -> fixed by adding `hicolor_icon_theme` to `ibus_wrapper`
- ~~sunpinyin:  lib, data, fcitx packaged separately. fcitx engine miss to find the dictionary file:
  looking for `/lib/sunpinyin/data/lm_sc.t3g` but data is in `/share/sunpinyin/lm_sc.t3g`~~ -> Splitting sunpinyin packaging in a future PR
### preparing a custom nixpkgs
1. get a copy of `nixpkgs`
   
   ```
   git clone https://github.com/NixOS/nixpkgs.git
   cd nixpkgs
   ```
2. get this PR
   
   ```
   git fetch origin pull/11254/head:standardize-im
   ```
3. add nixpgs-channel as a remote and get `nixos-unstable`
   
   ```
   git remote add channels https://github.com/NixOS/nixpkgs-channels.git
   git fetch channels nixos-unstable
   git checkout nixos-unstable
   ```
4. create a test branch based on `standardize-im`
   
   ```
   git checkout standardize-im
   git branch test/11254
   ```
5. rebase `test/11254`
   
   ```
   git rebase --onto nixos-unstable master test/11254
   ```
   
   note: A few conflicts in `nixos/modules/module-list.nix` might happen.
### Fcitx

in `/etc/nixos/configuration.nix` add:

```
  i18n.inputMethod = {
    enabled = "fcitx";
    fcitx.engines = with pkgs.fcitx-engines; [ hangul mozc anthy chewing table-other ];
  };
```

and rebuild using the custom nixpkgs:

```
nixos-rebuild -I nixpkgs=/PATH/TO/CUSTOM/NIXPKGS switch
```

After restarting the X server, fcitx should be active.
note: mozc might require a computer reboot to work.
### IBus

in `/etc/nixos/configuration.nix` add:

```
  i18n.inputMethod = {
    enabled = "ibus";
    ibus.engines = with pkgs.ibus-engines; [ hangul mozc anthy ];
  };
```
### Uim

in `/etc/nixos/configuration.nix` add:

```
  i18n.inputMethod.enabled = "uim";
```
### nabi

in `/etc/nixos/configuration.nix` add:

```
  i18n.inputMethod.enabled = "nabi";
```

and rebuild using the custom nixpkgs:

```
nixos-rebuild -I nixpkgs=/PATH/TO/CUSTOM/NIXPKGS switch
```

After restarting the X server, ibus should be active.
note: mozc might require a computer reboot to work.
## Testing

Each input method and engine combination should be tested on as many applications toolkits as possible.

sample applications:
- Gtk2: `geany`
- Gtk3: `gnome3.gedit`
- Qt4: `kde4.kate`
- Qt5: `quasselClient_qt5`, `kde5.kate`
  - pyqt5: `qutebrowser`
- Elementary: `e19.terminology`
- wxWidgets: `filezilla`
- Java: `jedit`
### Test results

| IM | GTK2 | GTK3 | Qt4 | Qt5 | PyQt5 | Elementary | wxWidgets | Java |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Fcitx | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ |
| Ibus | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ | ◯ |
| uim | ◯ | ◯ | ◯ | -- | -- | ◯ | ◯ | ◯ |
| nabi | ◯ | ◯ | ◯ | -- | -- | ◯ | ◯ | ◯ |

◯ working
X not working
-- not supported
#### Fcitx PyQt5

~~`fcitx-qt5` is using `qt55` but `pyqt5` is using `qt54`, will be fixed when `pyqt5` will be updated to `qt55`~~ -> Fixed
# Possible improvements
- The `ibus-wrapper` and `fcitx-wrapper` code (taken from pidgin plugins) feels like it could be improved.
# Updates
- 10 Feb: added ibus related commits
- 13 Feb fixed ibus icon issues, updated the testing section
- 17 Feb: added uim and nabi commits  
              resolved conflicts with release-notes
- 19 Feb: fixed typo in description (configure.nix -> configuration.nix)  
              removed module rename for `programs.ibus.enable`
- 20 Feb: added `fcitx-m17n` and `ibus-m17n`, updated the changelog accordingly

### CODE REVIEW

**Comment by iyzsong** [comment]
the `preInstall` hook of fcitx-qt5:
```
    substituteInPlace platforminputcontext/cmake_install.cmake \
       --replace ${qt5.base} $out
```
fails for me.

it seems that in the generated cmake_install.cmake file,
a qtbase path refered by `extra-cmake-modules` is used,
but it is different with qt5.base from buildInputs.

And:
```
$ nix-store -qR /nix/store/*extra-cmake-module*/ | grep qtbase
 /nix/store/52zc0n2m872yjr2zs6janc6v26ggb62l-qtbase-5.5.1
$ nix-build -A qt5.base`, show:
 /nix/store/h11crcgkl89v3pixrdbp65syc3qpz2na-qtbase-5.4.2
```
@ttuegel , how to get the qtbase that ECM using?

**Comment by ericsagnes** [comment]
@iyzsong Thanks for checking!
Does explicitly using qt 5.4 helps?

```diff
--- a/pkgs/tools/inputmethods/fcitx/fcitx-qt5.nix
+++ b/pkgs/tools/inputmethods/fcitx/fcitx-qt5.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, cmake, fcitx, kde5, qt5 }:
+{ stdenv, fetchurl, cmake, fcitx, kde5, qt54 }:
 
 stdenv.mkDerivation rec {
   name = "fcitx-qt5-${version}";
@@ -9,11 +9,11 @@ stdenv.mkDerivation rec {
     sha256 = "070dlmwkim7sg0xwxfcbb46li1jk8yd3rmj0j5fkmgyr12044aml";
   };
 
-  buildInputs = [ cmake fcitx kde5.extra-cmake-modules qt5.base ];
+  buildInputs = [ cmake fcitx kde5.extra-cmake-modules qt54.base ];
 
   preInstall = ''
     substituteInPlace platforminputcontext/cmake_install.cmake \
-      --replace ${qt5.base} $out
+      --replace ${qt54.base} $out
   '';
 
   meta = with stdenv.lib; {
```

**Comment by ttuegel** [comment]
@iyzsong The package needs to be upgraded to Qt 5.5; [the manual](http://nixos.org/nixpkgs/manual/#sec-language-qt) describes how Qt packaging has changed. If your package requires `extra-cmake-modules`, you should add it to [`kf5PackagesFun`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L6792) if it provides a library or use `kf5_stable.callPackage` instead of `callPackage` if it is simply an application; this will ensure that the package expression is provided a consistent set of dependencies.

**Comment by ericsagnes** [comment]
@iyzsong I updated the commit, could you try again? 
The new path of `fcitx-qt5` is `kf5Packages.fcitx-qt5`.

@ttuegel Thanks for the information. The updated module works fine with `kf5Packages.quassel_client` but fails with `qutebrowser` that is using `pyqt5`. Am I missing something in the derivation?

**Comment by iyzsong** [comment]
ah, cool!  I have to do my homework though :-)

**Comment by iyzsong** [comment]
I have cherry-pick 3 commit to master, hope it's ok.

**Comment by ericsagnes** [comment]
@iyzsong As I tested so far, there are some issues with fcitx and qutebrowser that is a qt5 application using `pyqt5`.  On my dev env, I have some problems with uim and Qt applications.
Toolsets modules seems a little tricky to make work and are quite tedious to test.

I would suggest to test each input method and engine on a different set of applications (Qt, GTK, ..), and if possible gather some feedback from multiple users to be sure it is working before commiting to master.

File: nixos/modules/i18n/inputMethod/fcitx.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+in
+{
+  options = {
+
+    i18n.inputMethod.fcitx = {
+      enable = mkOption {
+        type    = types.bool;
+        default = false;
+        example = true;
+        description = "Enable Fcitx input method";
```
Comment (nbp) [inline nixos/modules/i18n/inputMethod/fcitx.nix]: Please provide a better description of what `Fctix input method` is.

File: nixos/modules/i18n/inputMethod/fcitx.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      };
+      engines = mkOption {
+        type    = with types; listOf path;
+        default = [];
+        description = ''
```
Comment (nbp) [inline nixos/modules/i18n/inputMethod/fcitx.nix]: Add literal examples of the possible packages. (path?)

**Comment by laMudri** [comment]
I'm the maintainer of ibus-table and ibus-table-others. They currently don't do anything, and I plan on working on them over the next few weeks. If I manage to make something useful, how should I share it? Should I pretend that this PR doesn't exist and structure things in the old way, or do some IBus refactoring before starting?

**Comment by ericsagnes** [comment]
@laMudri Thanks for the input. This PR might take some take to get ready as it still require some work, tests and documentation. I hope we can get some more work done during this month Tokyo NixOS meetup.

I should be able to push an updated version of the PR based on current master with a few more commits for uim and ibus this week so you could use it and push your updates on that branch.

But as your work is more likely to be merged in master before this PR, using the current structure  and a dedicated PR should be easier and faster (and would just require to rebase this PR on the new master).

**Comment by ericsagnes** [comment]
@ttuegel I am stuck trying to make the `fcitx-qt5` plugin work with PyQt5 application (qutebrowser). 

I thought the problem was on the qutebrowser/pyqt5 side as `fcitx-qt5` works fine with normal Qt5 applications, so I tried to use the [wrapQtProgram](http://nixpaste.lbr.uno/uMY--Dl0) and calling the qutebrowser package with `qt5Libs.callPackage` but it didn't solve the issue.

What do you think could be the cause?

**Comment by ttuegel** [comment]
@ericsagnes Currently, PyQt5 in Nixpkgs is built with Qt 5.4, so it isn't going to be compatible with `fcitx-qt5` built with Qt 5.5. This is a pretty common problem with Qt plugins. Perhaps the maintainers (@svanderburg @iyzsong) can tell us if PyQt5 can be upgraded to Qt 5.5? It would be best if we could add it to `qt5LibsFun` so that there would always be a version of PyQt5 available for every version of Qt 5 in Nixpkgs.

**Comment by ericsagnes** [comment]
@ttuegel Thanks, I tried to update [pyqt](http://nixpaste.lbr.uno/JfMPcRpr) to use Qt 5.5 and the input method worked fine.
Just by curiosity, is there any special reason to have `qt5` defaulting to `qt54` while `qt5Libs` is defaulting to `qt55`?

**Comment by ttuegel** [comment]
@ericsagnes Most packages that depend on Qt 5 have not been updated to the new Qt infrastructure yet, so I can't automatically switch their Qt version. I'm waiting on the package maintainers, basically.

**Comment by ericsagnes** [comment]
@ttuegel The `fcitx` plugin for Qt5, `pkgs.kde5.fcitx-qt5`, was working on the previous `nixos-unstable`, but not with the most recent one that was released today. 
Do you know what could be the cause?

**Comment by ttuegel** [comment]
@ericsagnes I can't think of any major changes in Qt 5 in that time. What was the last revision you tested where it worked?

**Comment by ericsagnes** [comment]
@ttuegel My bad, after some investigation the problem was with the `fcitx` wrapper package. moving `fcitx-qt5` from the service to the wrapper fixed the problem.
Thanks for the fast reply!

**Comment by ericsagnes** [comment]
This PR is working and ready for review.
The test results are in the PR description, and there were no issues during the tests.
(all the input methods were tested with Japanese and Korean engines).

@lamudri I made a few trivial changes to `ibus-table` and `ibus-table-other`, could you check that they work as intended?

cc potential reviewers:
@gebner ibus maintainer
@iyzsong fcitx maintainer
@wavewave nabi maintainer
@taku0 uim main commiter
@nbp

**Comment by laMudri** [comment]
I believe that `ibus-table-others` shouldn't have the `meta.isIbusEngine` attribute, since it's intended as a plugin to `ibus-table`, not `ibus` directly. The stuff I'm working on [here](https://github.com/laMudri/nixpkgs/tree/ibus-table-new) better reflects that, but still isn't working.

Given that those packages don't work anyway, there isn't really anything to test. Does `ibus-anthy` need to be tested, though?

**Comment by ericsagnes** [comment]
update: There was a problem with ibus mozc `mozc.xml`, so I updated the `ibus-mozc: 2015-05-02 -> 2.17.2313.102` commit.

@laMudri if it doesn't need to be activated via `i18n.inputMethod.ibus.engines`,  `meta.isIbusEngine` is not needed. The `isIbusEngine` meta is just a way to limit the set of packages that can be used as `i18n.inputMethod.ibus.engines`. (for example `fcitx-table-other` needs `isFcitxEngine`)

I have tested `ibus` with `anthy` but it is good to test it on a different setup to be sure everything works fine.

**Comment by iyzsong** [comment]
Neat, thanks!

I'll do some tests on my laptop (non-NixOS) later.

**Comment by ericsagnes** [comment]
@iyzsong Thanks! Input methods are dependent on the environment variables and gtk/qt plugins set by the service files. This might be an issue on non-NixOS platforms.

**Comment by laMudri** [comment]
Can you give an example config? Whether or not I have
```
  i18n.inputMethod.ibus = {
    enable = true;
    plugins = [ pkgs.ibus-anthy ];
  };
```
I get
```
error: The option value `i18n.inputMethod.enabled' in `/home/james/ericsagnes-nixpkgs/nixos/modules/rename.nix' is not a null or one of "ibus", "fcitx", "nabi", "uim".
```
. I've tried actually setting that attribute (`i18n.inputMethod.enabled = "ibus";`), too.

**Comment by ericsagnes** [comment]
It looks like using `mkRenamedOptionModule` with different types was not a good idea.
I updated the PR to remove the following:

```
(mkRenamedOptionModule [ "programs" "ibus" "enable" ]  [ "i18n" "inputMethod" "enabled" ])
```

An example configuration for using ibus would be:

```
  i18n.inputMethod = {
    enabled = "ibus";
    ibus.engines = with pkgs.ibus-engines; [ hangul mozc anthy ];
  };
```

**Comment by ericsagnes** [comment]
update: I just saw that @astsmtl packaged `m17n` so I added the `fcitx-m17n` and `ibus-m17n` packages to the PR. Thank you @astsmtl!

So now, this PR should bring input support for most non-alphabet languages. 
Non alphabet input support being a must have feature in many countries, it would be great if we can have this PR merged for 16.03!
Please let me know if there is anything I can do to speed up the review/merge process.

File: nixos/doc/manual/release-notes/rl-unstable.xml [hunk 3]
```diff
@@ -197,6 +197,33 @@ fileSystems."/example" = {
...
+    names. The input method selection has been moved to a global
+    <literal>i18n.inputMethod.enabled</literal> option.
+    An example of the new style:
+
+<programlisting>
+i18n.inputMethod.enabled = "ibus";
+i18n.inputMethod.ibus.engines = with pkgs.ibus-engines; [ anthy mozc ];
+</programlisting>
+
+That is equivalent to the old version:
```
Comment (iyzsong) [inline nixos/doc/manual/release-notes/rl-unstable.xml]: There is no services for input method before, so I think just add them to the "new services" para should be enough.

**Comment by iyzsong** [comment]
Expect the ``release-note``, all commits look good to me, I think it's mergeable.
But I haven't actually test them (my setup is out of sync with master, low speed to hydra, etc.).

You should send an email to the ``nix-dev`` to draw more attention :-)

**Comment by ericsagnes** [comment]
@iyzsong Thanks for the review! I updated the release notes.

**Comment by laMudri** [comment]
IBus works on my machine. ibus-hangul doesn't (I mean, switching to it doesn't affect the input), but that's a different issue. I'll look into it.

**Comment by ericsagnes** [comment]
@laMudri `ibus-hangul` need some special setup to work.

1. add hangul to the ibus methods 
2. hit `alt+space` until hangul is active
3. hit `shift+space` to enable hangul mode

If you left click the hangul symbol in the systray, you should be able to change the shortcut for the hangul mode and other different settings. You can also launch the tool directly by running `ibus-setup-hangul`.

**Comment by laMudri** [comment]
Yep, all working, then.

**Comment by grahamc** [comment]
Another note is these lint issues:

```
22:06:18.262 > ibus-with-plugins: Lacks a maintainer
22:06:18.263 > ibus-with-plugins: Lacks a platform
22:06:18.263 > ibus-with-plugins: Lacks a license
22:06:18.263 > ibus-with-plugins: Lacks a description
```

Not sure if they need to be fixed, just a reminder.

**Comment by ericsagnes** [comment]
@grahamc Thanks for pointing. Yes, I raised some concerns in the issues description related `*-with-plugins` package. Actually the code is coming from `fcitx-with-plugins` that is in turn based on [`pidgin-with-plugins`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/pidgin/wrapper.nix). 
I created a dedicated issue ( #13156 ) about on this, as it has a wider scope than this PR.

**Comment by taku0** [comment]
Confirmed that uim (Anthy and SKK) works on Konsole (Qt4) and Firefox (GTK+2).

It does not work on Flash nor text fields on uim-pref-gtk and uim-pref-gtk3, but it seems not related to this PR because it does not work on stable branch neither.

**Comment by iyzsong** [comment]
I have test fcitx with various plugins, work fine.
Merged, thanks your work!

**Comment by ericsagnes** [comment]
Thanks for merging!

**Pull request merged**